### PR TITLE
Misc

### PR DESCRIPTION
**Pull request by jschueller**
- #1921 
- #1922 
- #1926 
- Avoid a numpy warning with the array interface as baseaddress yields None with an empty collection
- TBB is now a private dependency (we do not include tbb.h in public headers)
- Sample::exportCSV allows to tweak precision & format
- coupling_tools: rework api to specify stdout/stderr simultaneously, more similar to subprocess apis
- Add PlatformInfo::HasFeature, replacing all the IsAvailable methods

### CODE REVIEW

File: lib/src/Base/Stat/SampleImplementation.cxx [hunk 1]
```diff
@@ -2352,13 +2351,28 @@ void SampleImplementation::exportToCSVFile(const FileName & filename,
     }
     csvFile << "\n";
   }
+
+  const UnsignedInteger precision = ResourceMap::GetAsUnsignedInteger("Sample-CSVPrecision");
+  const String format = ResourceMap::Get("Sample-CSVFormat");
```
Comment (josephmure) [inline lib/src/Base/Stat/SampleImplementation.cxx:2356]: It is a good idea to allow users to tweak the precision and format, but couldn't they be specified as arguments (with default values given by ResourceMap) of `Sample::exportToCSVFile`?

Comment (jschueller) [reply lib/src/Base/Stat/SampleImplementation.cxx:2356]: that would be too much arguments to provide, these are for less frequentl use-cases (and we dont have keyword arguments)

Comment (jschueller) [reply lib/src/Base/Stat/SampleImplementation.cxx]: yes, we can have arguments

File: ChangeLog [hunk 2]
```diff
@@ -17,6 +17,7 @@
  * Removed deprecated MultiFORM.setMaximumNumberOfDesignPoints/getMaximumNumberOfDesignPoints
  * Removed deprecated SubsetSampling.getNumberOfSteps
  * coupling_tools.execute: deprecated get_stderr,get_stdout for capture_output, returns CompletedProcess
+ * Deprecated Nlopt|Ceres|Bonmin|CMinpack|Ipopt|TBB|HMatrixFactory::IsAvailable in favor of PlatformInfo::HasFeature
```
Comment (sofianehaddad) [inline ChangeLog:20]: Nice feature. Could we complete with a `PlatformInfo::GetFeatures`?

File: python/src/Sample_doc.i.in [hunk 3]
```diff
@@ -797,6 +797,12 @@ separator : str
     Default uses `csv-file-separator` from the :class:`~openturns.ResourceMap`.
 numSeparator : str, default='.'
     Decimal separator.
+precision : int
+    Numerical precision
+    Default takes `Sample-CSVPrecision` entry from :class:`~openturns.ResourceMap`.
+format : str, value is one of scientific/fixed/default
+    Floating-point formatting
+    Default takes `Sample-CSVFormat` entry from :class:`~openturns.ResourceMap`.
```
Comment (sofianehaddad) [inline python/src/Sample_doc.i.in:805]: A word on possible format?

**Review by sofianehaddad — COMMENTED** [review]
Why did you introduce `TBBImplementation`? I probably missed something ...

File: python/src/Sample_doc.i.in (hunk 3 above)
Comment (jschueller) [reply python/src/Sample_doc.i.in:805]: ok

File: ChangeLog (hunk 2 above)
Comment (jschueller) [reply ChangeLog:20]: ok

**Comment by jschueller** [comment]
@sofianehaddad good to go ?

**Review by josephmure — APPROVED** [review]
The parts about TBB I do not really understand, but what I can understand seems OK to me.

**Comment by jschueller** [comment]
The new tbb header just allows to hide the template details that depend on tbb.h, so that client code dont need to install the tbb development package, that's it.

**Comment by jschueller** [comment]
see also https://github.com/openturns/openturns/issues/1229

**Comment by josephmure** [comment]
That's a quick merge, I am the only reviewer who approved!

**Comment by josephmure** [comment]
Otherwise thanks for the info!

**Comment by jschueller** [comment]
> That's a quick merge, I am the only reviewer who approved!

sofiane was ok too

**Review by regislebrun — APPROVED** [review]

**Comment by regislebrun** [comment]
Wow! I was reviewing the PR when it has been merged! How do you manage to *know* that I will approve it ;-)?

**Comment by jschueller** [comment]
that settles it, oof :)

**Comment by jschueller** [comment]
I guess we're in the multiverse were that PR is good ^^

**Pull request merged**